include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -1752,656 +1561,192 @@ namespace internal
-                      "n_rows==n_columns or single-line interpolation"));
-    AssertIndexRange(direction, dim);
-    constexpr int mm = contract_over_rows ? n_rows : n_columns,
-                  nn = contract_over_rows ? n_columns : n_rows;
-
-    constexpr int stride    = Utilities::pow(n_columns, direction);
-    constexpr int n_blocks1 = one_line ? 1 : stride;
-
-    // The number of blocks depend on both direction and dimension.
-    constexpr int n_blocks2 =
-      (dim - direction - 1 == 0) ?
-        1 :
-        ((direction == normal_dir) ?
-           Utilities::pow((n_rows - 1),
-                          (direction >= dim) ? 0 : dim - direction - 1) :
-           (((direction < normal_dir) ? (n_rows + 1) : n_rows) *
-            ((dim - direction == 3) ? n_rows : 1)));
-
-    for (int i2 = 0; i2 < n_blocks2; ++i2)
-      {
-        for (int i1 = 0; i1 < n_blocks1; ++i1)
-          {
-            Number x[mm];
-            for (int i = 0; i < mm; ++i)
-              x[i] = in[stride * i];
-
-            for (int col = 0; col < nn; ++col)
-              {
-                Number2 val0;
-
-                if (contract_over_rows)
-                  val0 = shape_data[col];
-                else
-                  val0 = shape_data[col * n_columns];
-
-                Number res0 = val0 * x[0];
-                for (int i = 1; i < mm; ++i)
-                  {
-                    if (contract_over_rows)
-                      val0 = shape_data[i * n_columns + col];
-                    else
-                      val0 = shape_data[col * n_columns + i];
-
-                    res0 += val0 * x[i];
-                  }
-                if (add)
-                  out[stride * col] += res0;
-
-                else
-                  out[stride * col] = res0;
-              }
-
-            if (one_line == false)
-              {
-                ++in;
-                ++out;
-              }
-          }
-        if (one_line == false)
-          {
-            in += stride * (mm - 1);
-            out += stride * (nn - 1);
-          }
-      }
-  }
-
-  template <int dim,
-            int n_rows,
-            int n_columns,
-            int normal_dir,
-            typename Number,
-            typename Number2>
-  template <int  face_direction,
+  template <int  n_rows_template,
+            int  stride_template,
             bool contract_onto_face,
             bool add,
-            int  max_derivative>
-  inline void
-  EvaluatorTensorProductAnisotropic<
-    evaluate_raviart_thomas,
-    dim,
-    n_rows,
-    n_columns,
-    normal_dir,
-    Number,
-    Number2>::apply_face(const Number *DEAL_II_RESTRICT in,
-                         Number *DEAL_II_RESTRICT       out) const
+            int  max_derivative,
+            typename Number,
+            typename Number2>
+  inline std::enable_if_t<!contract_onto_face, void>
+  interpolate_to_face(const Number2            *shape_values,
```

Comment:
Why not skip the `std::enable_if_t` and call this function `interpolate_from_face` if we have separate functions depending on the template parameter anyway?

Reply:
I would like to do that, but the restriction is that we call them at the same site with just the template parameter distinguishing between the two. I think it is good to have separate functions rather then mix them in the same function with some additional `if` statements, but we could discuss. It would be nice if I could use some `using` statement that give me shortcuts for function names (like we can do for templated classes) and just change the template parameter that changes, but I could not figure that out.